src/backend/distributed/executor/executor.c

```diff
@@ -1346,6 +1407,40 @@ ConnectionStateMachine(WorkerSession *session)
 }
 
 
+/*
+ * Use2PCIfParticipatingNodesExpanded ensures that the execution uses
+ * 2PC if the transaction is expanded to access a new worker node.
+ */
+static void
+Use2PCIfParticipatingNodesExpanded(WorkerSession *session)
+{
+	DistributedExecution *execution = NULL;
+
+	if (DoesCoordinatedTransactionUse2PC() ||
+		MultiShardCommitProtocol != COMMIT_PROTOCOL_2PC)
+	{
+		/* already using or doesn't require 2PC */
+		return;
+	}
+
+	/*
+	 * If we're expanding the set nodes that participate in the distributed
+	 * transaction, conform to MultiShardCommitProtocol.
+	 */
+	execution = session->workerPool->distributedExecution;
+	if (execution->isTransaction &&
+		XactModificationLevel == XACT_MODIFICATION_DATA)
+	{
+		MultiConnection *connection = session->connection;
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+		if (transaction->transactionState == REMOTE_TRANS_INVALID)
```

Comment:
we may need to be careful here because we've started opening transaction blocks for SELECTs in a transaction block, so something like `BEGIN; SELECT * FROM test WHERE x = 2; INSERT INTO test VALUES (1,2); INSERT INTO test VALUES (2,1);` might not activate 2PC because the `transactionState` is not invalid (did not verify yet).

Reply:
I haven't verified that as well, but as far as I can see, we're resetting the transaction state after successful command executions [here](https://github.com/citusdata/citus/compare/unified_executor#diff-4b24978830320f290897555d61ec2af3R1462)

Reply:
Ok, I see what you mean now. So, with `ConnectionModifiedPlacement` we should be fine?